### PR TITLE
bpo-39812: Remove daemon threads in concurrent.futures

### PR DESCRIPTION
**Pull request by aeros**
Removes daemon threads from the Executor internals in concurrent.futures to allow for compatibility with subinterpreters (since they no longer support daemon threads with https://bugs.python.org/issue37266).

/cc @pitrou 

https://bugs.python.org/issue39812

### CODE REVIEW

File: Lib/threading.py
Comment (aeros) [inline Lib/threading.py:1355]: I specified "CPython internal" instead of just "internal" to make it more clear that it's intended to be used across stdlib modules (as opposed to an internal helper function for `threading`).

At some point, I think this *could* graduate to the public API, but at the moment it's a bit too niche since subinterpreters have not yet officially made it into the stdlib. If such a utility is requested by users, I think it could be moved without much hassle though. That's partly why I included a detailed docstring.

File: Lib/concurrent/futures/process.py
Comment (aeros) [inline Lib/concurrent/futures/process.py:736]: This change is just to make the atexit registration be in the same location for both ProcessPoolExecutor and ThreadPoolExecutor (right under the definition for `_python_exit()`). It can be removed if needed, but I think it will make them easier to locate for those reading over the implementation details.

**Comment by aeros** [comment]
I considered adding a unit test for `_threading_atexit()` that would attempt to register a mock object, and ensure it was called once after thread shutdown process is complete. However, upon trying to implement it, I think it might be a bit more trouble than it's worth for a fairly simple internal function.

That being said, I think it would be worthwhile to add a test along those lines if it were to be added to the public API for `threading`.

File: Lib/concurrent/futures/process.py
Comment (aeros) [inline Lib/concurrent/futures/process.py:74]: I believe this section is no longer relevant now that the workers are no longer daemon threads. Correct me if I'm mistaken.

**Comment by aeros** [comment]
(Note that `AMD64 Windows7 SP1 PR` and `x86 Windows7 PR` can be safely disregarded since Windows 7 isn't supported on 3.x)

**Comment by aeros** [comment]
I'm 99.99% certain the `test_socket` failure in `FreeBSD Shared PR` is unrelated to the PR, as it has been occurring intermittently in [recent commits to master](https://github.com/python/cpython/commits/master) as well. Also, it's further reinforced by the fact that this PR doesn't affect the `socket` module.

```
1 re-run test:
    test_socket
Total duration: 33 min 55 sec
Tests result: FAILURE then FAILURE
sys:1: ResourceWarning: unclosed <socket.socket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=132, laddr=('127.0.0.1', 51870)>
ResourceWarning: Enable tracemalloc to get the object allocation traceback
sys:1: ResourceWarning: unclosed <socket.socket fd=4, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=132, laddr=('127.0.0.1', 52586)>
ResourceWarning: Enable tracemalloc to get the object allocation traceback
sys:1: ResourceWarning: unclosed <socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=132, laddr=('127.0.0.1', 49295)>
ResourceWarning: Enable tracemalloc to get the object allocation traceback
*** Error code 2
```

**Comment by aeros** [comment]
I opened a separate issue to address the above (unrelated) refleaks at https://bugs.python.org/issue40061.

File: Lib/threading.py
Comment (pitrou) [inline Lib/threading.py]: I don't think there is any downside to assigning an empty list directly here?

File: Lib/threading.py
Comment (pitrou) [reply Lib/threading.py:1355]: This sounds good to me, thanks for the explanation.

File: Lib/threading.py
Comment (pitrou) [inline Lib/threading.py]: I wonder if it's worth checking if one of the atexit callbacks didn't call `_register_atexit`?

**Review by pitrou — COMMENTED** [review]
Thanks for doing this. I think `_register_atexit` deserves at least some basic unit test(s). It should be fairly easy by spawning e.g. a subprocess and checking its output.

File: Lib/threading.py
Comment (aeros) [reply Lib/threading.py]: That's a good point that I hadn't considered. I personally don't see a strong realistic use case for having to register another atexit callback with an atexit callback, so I think that can be unsupported for now (until someone comes up a good use).

I'm not certain what the behavior should be though. Would it simply be ignored, raise an exception, or something else? I'm leaning towards an exception (to be explicit), but I could see that causing issues with interupting the shutdown process by not allowing the other atexits to be called.

Perhaps there could be a global flag that gets set to true in `_shutdown()` such as `_SHUTTING_DOWN`, which would be checked in `_register_atexit()` prior to registering the callback. This would ensure that new callbacks couldn't be registered once the shutdown process starts.

File: Lib/threading.py
Comment (aeros) [reply Lib/threading.py]: I did this so that a list is not spawned every time `threading` is imported, even when `_register_atexit()` is not used. It's not very substantial, but I've always made the habit of avoiding as much overhead as possible in module-level globals. IMO, it's one of those practices that are individually insignificant, but can add up over time.

That being said, I would be okay with just initializing the list directly here, since the overhead for creating a single list is so small. I'll leave it up to you, as the primary maintainer for `threading`.

**Comment by aeros** [comment]
> Thanks for doing this.

No problem, thanks for the review. :-)

> I think _register_atexit deserves at least some basic unit test(s). It should be fairly easy by spawning e.g. a subprocess and checking its output.

Yeah, after having some additional time to think about it, I think we'd be better off with a couple of basic unit tests. I'll work on those next.

File: Lib/threading.py
Comment (pitrou) [reply Lib/threading.py]: Yes, you could e.g. raise an exception there.

File: Lib/threading.py
Comment (pitrou) [reply Lib/threading.py]: I think initializing the list here is simpler. The overhead of a single (empty) list object is tiny.

File: Lib/threading.py
Comment (aeros) [reply Lib/threading.py]: Done.

**Comment by aeros** [comment]
Going to restart the CI (via close and re-open) since it failed in the "Install Dependencies" stage.

**Comment by aeros** [comment]
@pitrou I believe that I've addressed the review comments. Let me know if the unit tests I added are sufficient.

**Comment by aeros** [comment]
Also, would it be worth adding a "What's New" entry to mention that the executors no longer rely on daemon threads? It's not exactly a new feature and was mainly an internal change, but I think it would be useful for users to know that it improves:

1) Predictability by joining the threads prior to interpreter shutdown (which has been a long-standing issue, AFAIK)
2) Compatibility with subinterpreters

**Comment by pitrou** [comment]
(closing and reopening to trigger CI)

**Comment by pitrou** [comment]
> Also, would it be worth adding a "What's New" entry to mention that the executors no longer rely on daemon threads?

That's a good idea.

**Comment by pitrou** [comment]
Oh, and yes, the unit tests look sufficient for an internal function :-)

**Comment by aeros** [comment]
@pitrou 
> (closing and reopening to trigger CI)

By the way, with GitHub actions for some reason, it leaves the previous CI failures and duplicates the tests when you close and re-open the PR. So the current 2 failures showing are from the first time (before I closed and re-opened it), that's why there's 3 of each GitHub actions test.

> That's a good idea.

Great! I'll work on the "What's New" entry next; it shouldn't take long. :-)

File: Doc/whatsnew/3.9.rst
Comment (aeros) [inline Doc/whatsnew/3.9.rst:199]: Using a tilde `~` shortens the Sphinx markup link to appear as just the last part, I.E. `ThreadPoolExecutor` and `ProcessPoolExecutor`. I typically use this when the classes are easily distinguishable based on their names alone, and the section already makes it clear what module it's in. This helps to make it a bit more succinct.

**Review by pitrou — APPROVED** [review]
Thanks for doing this @aeros !

File: Lib/concurrent/futures/process.py
Comment (laike9m) [inline Lib/concurrent/futures/process.py:311]: I'm actually depending on this behavior and my code breaks in Python 3.9. Is there a way to set the threads to be daemon? The error I got:

```
Traceback (most recent call last):
  File "/Users/laike9m/.pyenv/versions/3.9.0/lib/python3.9/threading.py", line 950, in _bootstrap_inner
    self.run()
  File "/Users/laike9m/.pyenv/versions/3.9.0/lib/python3.9/threading.py", line 888, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/laike9m/.pyenv/versions/cyberbrain-dev@3.9.0/lib/python3.9/site-packages/grpc/_server.py", line 875, in _serve
    if not _process_event_and_continue(state, event):
  File "/Users/laike9m/.pyenv/versions/cyberbrain-dev@3.9.0/lib/python3.9/site-packages/grpc/_server.py", line 839, in _process_event_and_continue
    rpc_state, rpc_future = _handle_call(event, state.generic_handlers,
  File "/Users/laike9m/.pyenv/versions/cyberbrain-dev@3.9.0/lib/python3.9/site-packages/grpc/_server.py", line 749, in _handle_call
    return _handle_with_method_handler(rpc_event, method_handler,
  File "/Users/laike9m/.pyenv/versions/cyberbrain-dev@3.9.0/lib/python3.9/site-packages/grpc/_server.py", line 722, in _handle_with_method_handler
    return state, _handle_unary_stream(rpc_event, state,
  File "/Users/laike9m/.pyenv/versions/cyberbrain-dev@3.9.0/lib/python3.9/site-packages/grpc/_server.py", line 640, in _handle_unary_stream
    return thread_pool.submit(_stream_response_in_pool, rpc_event, state,
  File "/Users/laike9m/.pyenv/versions/3.9.0/lib/python3.9/concurrent/futures/thread.py", line 163, in submit
    raise RuntimeError('cannot schedule new futures after '
RuntimeError: cannot schedule new futures after interpreter shutdown
```

which I believe is relevant. @aeros

Comment (aeros) [reply Lib/concurrent/futures/process.py:311]: @laike9m Hmm, does your code specifically rely on submitting new jobs *after* `executor.shutdown()` gets called? If so, would it be possible to move the `executor.shutdown()` to a later point (optimally, you should be done with using the executor before calling `shutdown()`)?

That being said, there may be some reason to consider allowing users to opt-in to using daemon threads; particularly since the decision to prevent subinterpreters from using daemon threads was reverted (which was originally the main motivation for this PR, with the benefit of more predictable shutdown behavior).

Do you have any thoughts @pitrou?

Comment (pitrou) [reply Lib/concurrent/futures/process.py:311]: @laike9m What do you mean with "break"? Does the process not exit cleanly?

Comment (laike9m) [reply Lib/concurrent/futures/process.py:311]: Let me describe my use case.

I'm using gRPC which starts a server using `ThreadPoolExecutor`:
https://github.com/laike9m/Cyberbrain/blob/master/cyberbrain/rpc_server.py#L235

This server runs in [another thread](https://github.com/laike9m/Cyberbrain/blob/master/cyberbrain/tracer.py#L159) besides the main thread, and would wait using [`server.wait_for_termination()`](https://github.com/laike9m/Cyberbrain/blob/master/cyberbrain/rpc_server.py#L256) until users explicitly ctrl-c. The main thread would run and finish the work, clients may or may not connect to the server before the main thread end, and **the requirement is for the server to respond even after the main thread has finished**. In Python <=3.8 this works, but in 3.9 I got the above error.

I'm not 100% that the root cause is this change, but I scanned through all changes in concurrent.futures in 3.9, and this seems to be the most likely one.

Comment (pitrou) [reply Lib/concurrent/futures/process.py:311]: I see, then I think you need to define a signal handler that will be called when Ctrl-C is pressed, and that will ask the server to stop gracefully.

Comment (laike9m) [reply Lib/concurrent/futures/process.py:311]: emm, it's actually not about what happens when users press Ctrl+C, but the server is not able to serve requests before that...

```
server starts listening --> main thread finishes ---> requests come ------> Ctrl + C
                                                    (problem is here)
```

Comment (laike9m) [reply Lib/concurrent/futures/process.py:311]: I'll try to come up with a minimal example that can reproduce this error.

Comment (pitrou) [reply Lib/concurrent/futures/process.py:311]: Ah, then it's simpler: you should ask the server to finish _and_ wait for it to finish before you let the main thread exit.

Comment (laike9m) [reply Lib/concurrent/futures/process.py:311]: Sounds good, thanks.

**Comment by aeros** [comment]
@gvanrossum this PR needs to be reverted, but I no longer have my commit bit

**Comment by gpshead** [comment]
> This PR needs to be reverted, but I no longer have my commit bit

Please open a new issue with reasoning.

In general we do not _want_ daemon threads anywhere anymore. Their existence has turned out to be a misfeature as they mean you cannot safely finalize a Python runtime when such threads are present.

**Comment by Bluehorn** [comment]
I am also curious why @aeros stated that this needs reverting. It looks like a backwards incompatible change (which is why I ended up reading this ticket), but it looks like affected projects can easily cope.

Did run into this because I manipulate interpreter shutdown for a test suite, which at times got stuck due to some non-daemon thread "helpfully" started in some library package. That code is not used in production but for diagnostics, as the failure was intermittent and difficult to reproduce.

> In general we do not want daemon threads anywhere anymore.

I think nobody "wants" units of execution that are just killed without notice in whatever state they are. I'd just like to advise against dropping them without a replacement.

The use case I see is dealing with threads getting stuck in some native code. This came up in my career a few times and was always a giant PITA to solve (or, actually, work around). What I'd like is a way to detach a thread in a way that it may not come back from extension code and just silently stops. I am not sure if and how that could be implemented, though.

Sorry for the noise, got longer than anticipated.

**Comment by pitrou** [comment]
@Bluehorn Useful explanation, thank you. I think @gpshead 's statement "Please open a new issue with reasoning" still holds, though.

**Comment by gvanrossum** [comment]
> I am also curious why @aeros stated that this needs reverting.

The leading theory is that he lost control over his account. That account's permissions have been dropped.

**Pull request merged**
